### ISSUE TITLE
postgresql parallel builds fail on Darwin

### ISSUE THREAD

**Issue by endgame** [NixOS/nixpkgs#51093]
## Issue description

While trying to fix #50902 I was unable to build `postgresql`. The important lines from the build log ([full log](https://github.com/NixOS/nixpkgs/files/2617687/build.log)) seem to be:

```
ar crs libpgtypes.a numeric.o datetime.o common.o dt_common.o timestamp.o interval.o pgstrcasecmp.o
echo 'Libs: -L/nix/store/7l8df1na81psj9cx56yjs4m9084nrayr-postgresql-9.6.10-lib/lib -lecpg_compat' >>libecpg_compat.pc
error: /nix/store/p1zg3dnaaglqm34pq12ynfa2pc3r2lq8-cctools-port-895/bin/ranlib: can't open file: libpgtypes.a (No such file or directory)
ar: internal ranlib command failed
make[5]: *** [../../../../src/Makefile.shlib:306: libpgtypes.a] Error 1
```

The build succeeds with `nix-build --cores 1 --max-jobs 1 -E 'let nixpkgs = import (builtins.fetchTarball { url = "https://github.com/endgame/nixpkgs/archive/bc565bba2762edf12a40fb0c153d676e66e28faa.tar.gz"; sha256 = "1hdcg4b84c0mw959d6x9pxal95j58ks82fpfsy5sbazs6s40hdhy"; }) {}; in nixpkgs.haskellPackages.postgresql-libpq'`

### Steps to reproduce

This build may fail:
```
nix-build -E 'let nixpkgs = import (builtins.fetchTarball { url = "https://github.com/endgame/nixpkgs/archive/bc565bba2762edf12a40fb0c153d676e66e28faa.tar.gz"; sha256 = "1hdcg4b84c0mw959d6x9pxal95j58ks82fpfsy5sbazs6s40hdhy"; }) {}; in nixpkgs.haskellPackages.postgresql-libpq'
```

(While it points to a commit in my fork of nixpkgs, this fork does not change the `postgresql` package. It updates something that depends on `postgresql` which I still haven't got building on Darwin.)

## Technical details

`nix-info -m`:
```
 - system: `"x86_64-darwin"`
 - host os: `Darwin 17.7.0, macOS 10.13.6`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1.3`
 - channels(che347): `"nixpkgs-19.03pre159614.f52b7f5c345, nixpkgs-18.09-darwin-18.09pre152973.012b9e6c219"`
 - nixpkgs: `/Users/che347/.nix-defexpr/channels/nixpkgs`
```

**Comment by endgame** [comment]
https://logs.nix.samueldr.com/nixos/2018-04-16#1117127; seems to be related.

**Comment by alyssais** [comment]
Do we know if that failure in the IRC logs you linked was on Darwin? We could disable parallel building on Darwin, but if this affects Linux too we'd be better doing both.

**Comment by alyssais** [comment]
Doesn't hurt to propose it for Darwin now, I guess. Can always add Linux later.

**Comment by endgame** [comment]
I think it built correctly on my GNU/Linux machine, but since this is possibly a non-deterministic failure that's not strong evidence for anything in particular.

**Comment by veprbl** [comment]
It would be nice to have this reported upstream as well.

**Comment by endgame** [comment]
There now: https://www.postgresql.org/message-id/15525-a30da084f17a1faa%40postgresql.org

**Comment by alyssais** [comment]
@delroth my guess is that it's trying to open a file that hasn't been created yet, because of a incorrect dependency graph – if that were the case, would make sense that disabling parallel building would fix it.

> can't open file: libpgtypes.a (No such file or directory)

**Comment by endgame** [comment]
Agreed with your reading @delroth that is appears to happen within the `ar rcs`, but all I can say is that it was repeatedly failing to build, then passed with `nix-build --cores 1 --max-jobs 1`.

**Comment by delroth** [comment]
@alyssais but that's the thing -- ar should be the command creating the archive. I don't see how this could be a make parallelism problem if the same command (at the Makefile level) which creates the .a also fails to read it for one of its steps. This seems like papering over an issue that's not properly understood.

**Comment by alyssais** [comment]
I see what you mean. It could indeed be something different, and if we see it again after disabling parallel builds I'm more than happy to be reverted, but I think it's worth a shot, given what @endgame says.

**Comment by endgame** [comment]
I found https://seclists.org/nmap-dev/2009/q1/166 which seems to discuss a similar problem with races in ranlib invocations on ~~OSX~~ macOS.

I note that the build log calls `ar crs libpgtypes.a ...` multiple times. Is it possible that they are stepping over each other?

**Comment by delroth** [comment]
@endgame I think that's indeed what's happening. Starting in 9.1.0 (postgres/postgres@19e231bb) this is in the postgres codebase:

```
./src/interfaces/ecpg/compatlib/Makefile:       $(MAKE) -C $(top_builddir)/src/interfaces/ecpg/pgtypeslib all
./src/interfaces/ecpg/ecpglib/Makefile: $(MAKE) -C $(top_builddir)/src/interfaces/ecpg/pgtypeslib all
```

If both of these end up happening in parallel then they would race over libpgtypes.a. Can you pass this back to upstream? I can't easily reply to the existing thread...

**Comment by endgame** [comment]
Will do @delroth ; I need to reply to the thread there anyway.

**Comment by endgame** [comment]
Reported @delroth's findings upstream: https://www.postgresql.org/message-id/87k1kwlvad.fsf%40jackkelly.name
